examples/chrono/chrono.cpp

```diff
@@ -0,0 +1,219 @@
+﻿/*
+ * Copyright (c) 2022, Christopher Kormanyos
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#include <chrono>
+
+#if defined(__GNUC__) && defined(__AVR__)
+#define ATTRIBUTE(...) __attribute__((__VA_ARGS__))
+#else
+#define ATTRIBUTE(X)
+#endif
+
+namespace mcal {
+
+namespace reg {
+
+// Bit values.
+constexpr auto bval5      = static_cast<std::uint8_t>(UINT8_C(1) << 5U);
+
+constexpr auto sfr_offset = static_cast<std::uint8_t>(UINT8_C(0x20));
+
+// System registers.
+constexpr auto mcusr      = static_cast<std::uint8_t>(UINT8_C(0x14) + sfr_offset);
+
+// Port registers.
+constexpr auto pinb       = static_cast<std::uint8_t>(UINT8_C(0x03) + sfr_offset);
+constexpr auto ddrb       = static_cast<std::uint8_t>(UINT8_C(0x04) + sfr_offset);
+constexpr auto portb      = static_cast<std::uint8_t>(UINT8_C(0x05) + sfr_offset);
+
+// Timer registers
+constexpr auto tifr0      = static_cast<std::uint8_t>(UINT8_C(0x15) + sfr_offset);
+constexpr auto timsk0     = static_cast<std::uint8_t>(UINT8_C(0x6E));
+constexpr auto tccr0b     = static_cast<std::uint8_t>(UINT8_C(0x25) + sfr_offset);
+constexpr auto tcnt0      = static_cast<std::uint8_t>(UINT8_C(0x26) + sfr_offset);
+
+// Watchdog registers
+constexpr auto wdtcsr     = static_cast<std::uint8_t>(UINT8_C(0x60));
+
+} // namespace reg
+
+namespace gpt {
+
+using value_type = std::uint64_t;
+
+auto init() -> void;
+
+auto get_time_elapsed() -> value_type;
+
+} // namespace gpt
+
+namespace wdg {
+
+auto init() -> void;
+
+inline void trigger()
+{
+  #if defined(__GNUC__) && defined(__AVR__)
+  asm volatile("wdr");
+  #endif
+}
+
+} // namespace gpt
+} // namespace mcal
+
+namespace local {
+
+mcal::gpt::value_type mcal_gpt_system_tick;
+
+using clock_type = std::chrono::steady_clock;
+
+auto delay_1s() noexcept -> void;
+
+} // namespace local
+
+auto main(void) -> int
+{
+  // Initialize the microcontroller abstraction layer.
+  mcal::wdg::init();
+  mcal::gpt::init();
+
+  // Enable all interrupts.
+  #if defined(__GNUC__) && defined(__AVR__)
+  asm volatile("sei");
+  #endif
+
+  using port_type = std::uint8_t;
+
+  // Set the port pin value to low.
+  *reinterpret_cast<volatile port_type*>(mcal::reg::portb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) & static_cast<port_type>(~mcal::reg::bval5));
+
+  // Set the port pin direction to output.
+  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) | mcal::reg::bval5);
```

Comment:
```suggestion
  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) | mcal::reg::bval5);
```

but I don't see the benefit (for a _chrono_ example) over

```suggestion
  DDRB |= (1<<5u);
```

This would have allowed the error to be seen much more quickly.

Reply:
> have allowed the error to be seen much more quickly

I was not aware of an error at this time. Did you detect one?

Reply:
Handled this with use of _regular_ AVR register-access tyles and idioms via `<avr/io.h>`.

Reply:
> I was not aware of an error at this time. Did you detect one?

Your code is equivalent to `DDRB = PORTB | (1<<5)`, which does not make sense. You probably meant `DDRB = DDRB | (1<<5)`.

Reply:
> Your code is equivalent to `DDRB = PORTB | (1<<5)`, which does not make sense. You probably meant 

Ugghhh... Sigh... I do that kind of thing all the time.

Well, I believe the entire line was (as below) which _may_ have been correct --- albeit a long line... This style is a comparatively safe, reusable tech for a `volatile` read-modify-write direct-memory-access sequence --- but it's non-atomic, maybe a bit slow.

```C
  // Set the port pin direction to output.
  *reinterpret_cast<volatile port_type*>(mcal::reg::ddrb) = static_cast<port_type>(*reinterpret_cast<volatile port_type*>(mcal::reg::portb) | mcal::reg::bval5);
```

Thanks. Good catch. Could you please also check the revised edition?

These things do get hard to read. And, moving forward, those AVR macros do a neat out-of-sight-hiding of the real direct-memroy-access going on, so I'm cool with either way...